File: product-processor-service/src/main/java/com/example/product_processor_service/configuration/kafka/factory/listener/KafkaListenerContainerFactoryConfiguration.java
package com.example.product_processor_service.configuration.kafka.factory.listener;

import com.example.product_processor_service.model.product.ProductPublisherDTO;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.config.KafkaListenerContainerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.listener.ConcurrentMessageListenerContainer;

@Configuration
public class KafkaListenerContainerFactoryConfiguration {
    @Bean
    public KafkaListenerContainerFactory<ConcurrentMessageListenerContainer<String, ProductPublisherDTO>> productPublisherListenerContainerFactory(ConsumerFactory<String, ProductPublisherDTO> consumerFactory) {
        var factory = new ConcurrentKafkaListenerContainerFactory<String, ProductPublisherDTO>();
        factory.setConsumerFactory(consumerFactory);

        return factory;
    }

    @Bean
    public KafkaListenerContainerFactory<ConcurrentMessageListenerContainer<String, String>> stringListenerContainerFactory(ConsumerFactory<String, String> consumerFactory) {
        var factory = new ConcurrentKafkaListenerContainerFactory<String, String>();
        factory.setConsumerFactory(consumerFactory);

        return factory;
    }
}

File: email-sender-service/src/main/kotlin/com/example/email_sender_service/configuration/kafka/factory/listener/KafkaListenerContainerFactoryConfiguration.kt
package com.example.email_sender_service.configuration.kafka.factory.listener

import com.example.email_sender_service.model.mail.dto.MailMessage
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.config.KafkaListenerContainerFactory
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.listener.ConcurrentMessageListenerContainer

@Configuration
class KafkaListenerContainerFactoryConfiguration {
    @Bean
    fun mailMessageListenerContainerFactory(consumerFactory: ConsumerFactory<String, MailMessage>): KafkaListenerContainerFactory<ConcurrentMessageListenerContainer<String, MailMessage>> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, MailMessage>()
        factory.consumerFactory = consumerFactory

        return factory
    }
}